File: frontend/src/app/types/pipeline.ts
export type StageStatus =
  | "idle"
  | "running"
  | "completed"
  | "failed"
  | "awaiting_review"
  | "rejected";

export interface PipelineStage {
  id: string;
  name: string;
  nameEn: string;
  agent: string;
  status: StageStatus;
  duration?: number;
  tokens?: number;
  output?: string;
  isCheckpoint?: boolean;
  startedAt?: string;
  completedAt?: string;
}

export interface PipelineGitStageCommit {
  stage_type: string;
  commit_sha: string;
  commit_message: string;
  committed_at: string;
  files_changed: string[];
}

export interface PipelineGitContext {
  mode: "disabled" | "worktree";
  enabled: boolean;
  repo_root?: string | null;
  base_branch?: string | null;
  base_commit?: string | null;
  worktree_path?: string | null;
  working_branch?: string | null;
  initialized: boolean;
  stage_commits: PipelineGitStageCommit[];
  total_files_changed: string[];
  head_commit?: string | null;
  diff_stats?: {
    files?: number;
    insertions?: number;
    deletions?: number;
  } | null;
  pr_title?: string | null;
  pr_description?: string | null;
  pr_command?: string | null;
}

export interface Pipeline {
  id: string;
  name: string;
  description: string;
  status: "running" | "completed" | "failed" | "paused" | "pending" | "cancelled";
  progress: number;
  currentStage: number;
  stages: PipelineStage[];
  createdAt: string;
  updatedAt: string;
  template?: string;
  projectPath?: string;
  projectSummary?: string;
  requirementDocPath?: string;
  solutionDocPath?: string;
}

export interface CreatePipelinePayload {
  projectPath: string;
  requirement: string;
}

File: backend/src/models/pipeline.py
"""FlowState pipeline 数据模型"""

from enum import Enum
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field, ConfigDict


class PipelineStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    PAUSED = "paused"
    WAITING_HUMAN = "waiting_human"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class StageType(str, Enum):
    REQUIREMENT = "requirement_analysis"
    SOLUTION = "solution_design"
    CODING = "coding"
    TESTING = "testing"
    REVIEW = "code_review"
    DELIVERY = "delivery"


class StageStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    WAITING_HUMAN = "waiting_human"
    APPROVED = "approved"
    REJECTED = "rejected"
    COMPLETED = "completed"
    FAILED = "failed"


class ApproveAction(str, Enum):
    APPROVE = "approve"
    REJECT = "reject"
    REVISE = "revise"


class GitMode(str, Enum):
    DISABLED = "disabled"
    WORKTREE = "worktree"


class StageCommit(BaseModel):
    """每个 stage 完成时的 commit 锚点。"""

    stage_type: StageType
    commit_sha: str
    commit_message: str
    committed_at: datetime
    files_changed: list[str] = Field(default_factory=list)


class GitContext(BaseModel):
    mode: GitMode = GitMode.DISABLED
    enabled: bool = False

    repo_root: Optional[str] = None
    base_branch: Optional[str] = None
    base_commit: Optional[str] = None

    worktree_path: Optional[str] = None
    working_branch: Optional[str] = None
    initialized: bool = False

    stage_commits: list[StageCommit] = Field(default_factory=list)

    total_files_changed: list[str] = Field(default_factory=list)
    head_commit: Optional[str] = None
    diff_stats: Optional[dict] = None

    pr_title: Optional[str] = None
    pr_description: Optional[str] = None
    pr_command: Optional[str] = None


class PipelineContext(BaseModel):
    """贯穿整个 Pipeline 的上下文"""
    project_path: str = ""
    project_summary: Optional[str] = None
    requirement_raw: str = ""
    requirement_doc: Optional[str] = None
    requirement_doc_path: Optional[str] = None
    solution_doc: Optional[str] = None
    solution_doc_path: Optional[str] = None
    solution_structured: Optional[Dict[str, Any]] = None
    generated_code: Optional[Dict[str, str]] = None
    test_report: Optional[str] = None
    review_report: Optional[str] = None
    delivery_result: Optional[str] = None
    code_diff: Optional[str] = None
    git: GitContext = Field(default_factory=GitContext)


class StageNode(BaseModel):
    """流水线中一个阶段节点"""
    model_config = ConfigDict(protected_namespaces=())

    stage_type: StageType
    status: StageStatus = StageStatus.PENDING
    agent_output: Optional[Dict[str, Any]] = None
    prompt_tokens: Optional[int] = None
    completion_tokens: Optional[int] = None
    total_tokens: Optional[int] = None
    model_name: Optional[str] = None
    human_feedback: Optional[str] = None
    human_approval: Optional[ApproveAction] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    retry_count: int = 0
    max_retries: int = 3


class Pipeline(BaseModel):
    """完整研发流水线"""
    id: str = Field(
        default_factory=lambda: f"pipe_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
    )
    title: str = ""
    status: PipelineStatus = PipelineStatus.PENDING
    stages: List[StageNode] = Field(default_factory=list)
    context: PipelineContext = Field(default_factory=PipelineContext)
    logs: List[str] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: datetime = Field(default_factory=datetime.now)
    error: Optional[str] = None
